keps/sig-storage/3314-csi-changed-block-tracking/README.md

```diff
@@ -0,0 +1,1692 @@
+  [TokenReview API](https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-review-v1/)
+  in order for its
+  [Kubernetes SnapshotMetadata](#the-kubernetes-snapshotmetadata-service-api)
+  gRPC service to validate the authentication token.
+
+- The CSI driver must be granted permission to use the Kubernetes
+  [SubjectAccessReview API](https://kubernetes.io/docs/reference/kubernetes-api/authorization-resources/subject-access-review-v1/)
+  in order for its
+  [Kubernetes SnapshotMetadata](#the-kubernetes-snapshotmetadata-service-api)
+  gRPC service to validate that a security token authorizes access to
+  VolumeSnapshot objects in a namespace.
+
+- The CSI driver presumably already has access rights to the VolumeSnapshot
+  and VolumeSnapshot content objects as they are within its purview.
+  This is needed for its
+  [Kubernetes SnapshotMetadata](#the-kubernetes-snapshotmetadata-service-api)
+  gRPC service.
+
+The proposal does not specify how such a security policy is to be configured.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### The CSI SnapshotMetadata Service API
+
+> In this section we use the terminology of the gRPC specification where the
+> word *service* is assumed to be a *gRPC service* and not a Kubernetes service,
+> while the word *plugin* is
+> used to refer to the software component that implements the gRPC service.
+
+The CSI specification will be extended with the addition of the following new, optional
+**SnapshotMetadata** [gRPC service](https://grpc.io/docs/what-is-grpc/core-concepts/#service-definition).
+The [SP Snapshot Metadata](#the-sp-snapshot-metadata-service) plugin implements
+this service.
+
+The service is defined as follows, and will be described in the sub-sections below.
+Refer to [CSI PR 551](https://github.com/container-storage-interface/spec/pull/551)
+for the official specification. 
+```
+service SnapshotMetadata {
+  rpc GetMetadataAllocated(GetMetadataAllocatedRequest)
+    returns (stream GetMetadataAllocatedResponse) {}
+  rpc GetMetadataDelta(GetMetadataDeltaRequest)
+    returns (stream GetMetadataDeltaResponse) {}
+}
+
+enum BlockMetadataType {
+  UNKNOWN=0;
+  FIXED_LENGTH=1;
+  VARIABLE_LENGTH=2;
+}
+
+message BlockMetadata {
+  int64 byte_offset = 1;
+  int64 size_bytes = 2;
+}
+
+message GetMetadataAllocatedRequest {
+  string snapshot_id = 1;
+  int64 starting_offset = 2;
+  int32 max_results = 3;
+  map<string, string> secrets = 4;
+}
+
+
+message GetMetadataAllocatedResponse {
+  BlockMetadataType block_metadata_type = 1;
+  int64 volume_capacity_bytes = 2;
+  repeated BlockMetadata block_metadata = 3;
+}
+
+message GetMetadataDeltaRequest {
```

Comment:
Is there a difference between `GetMetadataAllocatedRequest` and `GetMetadataDeltaRequest` aside from the `snapshot_id` vs `base_snapshot_id`/`target_snapshot_id` fields?

If there is no difference, maybe make base_snapshot_id optional (like the CSI spec's `Snapshot` `group_snapshot_id` field) and unify the two requests to reduce duplication.

Reply:
> Is there a difference between GetMetadataAllocatedRequest and GetMetadataDeltaRequest aside from the snapshot_id vs base_snapshot_id/target_snapshot_id fields?

No, not really. We had considered your suggestion and decided to explicitly document separate methods for clarity.